service/src/main/java/uk/gov/hmcts/reform/fpl/service/cmo/UploadCMOService.java

```diff
@@ -0,0 +1,164 @@
+
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toList;
+import static uk.gov.hmcts.reform.fpl.enums.CMOStatus.RETURNED;
+import static uk.gov.hmcts.reform.fpl.utils.DateFormatterHelper.DATE;
+import static uk.gov.hmcts.reform.fpl.utils.ElementUtils.asDynamicList;
+import static uk.gov.hmcts.reform.fpl.utils.ElementUtils.findElement;
+import static uk.gov.hmcts.reform.fpl.utils.JudgeAndLegalAdvisorHelper.formatJudgeTitleAndName;
+
+@Service
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class UploadCMOService {
+    private static final String CMO_HEARING_INFO_FIELD = "cmoHearingInfo";
+    private static final String CMO_JUDGE_INFO_FIELD = "cmoJudgeInfo";
+    private static final String PAST_HEARING_LIST_FIELD = "pastHearingList";
+
+    private static final String SINGLE = "SINGLE";
+    private static final String MULTI = "MULTI";
+    private static final String NONE = "NONE";
+    private static final String NUM_HEARINGS_FIELD = "numHearings";
+
+    private final ObjectMapper mapper;
+
+    public Map<String, Object> getInitialPageData(List<Element<HearingBooking>> hearings,
+                                                  List<Element<CaseManagementOrder>> unsealedOrders) {
+
+        List<Element<HearingBooking>> hearingsWithoutCMOs = getHearingsWithoutCMO(hearings, unsealedOrders);
+
+        Map<String, Object> data = new HashMap<>();
+        String textAreaKey = null;
+        String numHearings;
+        String showTextAreaKey = null;
+
+        switch (hearingsWithoutCMOs.size()) {
+            case 0:
+                numHearings = NONE;
+                break;
+            case 1:
+                numHearings = SINGLE;
+                textAreaKey = "singleHearingsWithCMOs";
+                showTextAreaKey = "showHearingsSingleTextArea";
+                data.putAll(getJudgeAndHearingDetailsSingle(hearingsWithoutCMOs.get(0).getId(), hearingsWithoutCMOs));
+                break;
+            default:
+                numHearings = MULTI;
+                textAreaKey = "multiHearingsWithCMOs";
+                showTextAreaKey = "showHearingsMultiTextArea";
+                data.put(PAST_HEARING_LIST_FIELD, buildDynamicList(hearingsWithoutCMOs));
+        }
+
+        String textAreaContent = buildHearingsWithCMOsText(unsealedOrders, hearings);
+
+        if (textAreaContent.length() != 0 && textAreaKey != null) {
+            data.put(textAreaKey, textAreaContent);
+            data.put(showTextAreaKey, "YES");
+        }
+
+        data.put(NUM_HEARINGS_FIELD, numHearings);
+
+        return data;
+    }
+
+    public List<Element<HearingBooking>> getHearingsWithoutCMO(List<Element<HearingBooking>> hearings,
+                                                               List<Element<CaseManagementOrder>> unsealedOrders) {
+        return hearings.stream()
+            .filter(hearing -> associatedToReturnedCMO(hearing, unsealedOrders)
```

Comment:
Is returning part of this story or just preparing for the future 🤔

Reply:
it wasn't part of the story but if they return a CMO then the hearing wouldn't have been picked up in the list again and would have remained stuck in the draft cmo list

Reply:
It wasn't too much effort, just adding some extra filtering and some tests to cover the scenario so I just did it

Reply:
👍